### ISSUE TITLE
JPA Server 5.0.1 - Validation of resources with extensions fails

### ISSUE THREAD

**Issue by j-fbriere** [jamesagnew/hapi-fhir#1860]
**Description:**
The $validate of a resource with extensions fails with errors when it shouldn't fail. Tests made on the JPA Server 5.0.1 (FHIR Version = R4).

**Steps to reproduce the behavior:**
 - Validate a ressource with the [elementdefinition-identifier](https://www.hl7.org/fhir/extension-elementdefinition-identifier.html) HL7 extension:

POST /Patient/$validate
```
{
    "resourceType": "Patient",
    "text": {
        "status": "generated",
        "div": "<div xmlns=\"http://www.w3.org/1999/xhtml\">Patient with extension</div>"
    },
    "extension": [
        {
            "url": "http://hl7.org/fhir/StructureDefinition/elementdefinition-identifier",
            "valueIdentifier": {
                "use": "temp",
                "system": "http://myIdentifier.org",
                "value": "12345678"
            }
        }
    ],
    "name": [
        {
            "family": "Patient with extension"
        }
    ]
}
```
Result:
```
{
  "resourceType": "OperationOutcome",
  "text": {
    "status": "generated",
    "div": "<div xmlns=\"http://www.w3.org/1999/xhtml\"><h1>Operation Outcome</h1><table border=\"0\"><tr><td style=\"font-weight: bold;\">ERROR</td><td>[Patient, Line 1, Col 2]</td><td><pre>The extension http://hl7.org/fhir/StructureDefinition/elementdefinition-identifier is not allowed to be used at this point (allowed = e:ElementDefinition; this element is [[Patient])</pre></td>\r\n\t\t\t\t\t\r\n\t\t\t\t\r\n\t\t\t</tr>\r\n\t\t</table>\r\n\t</div>"
  },
  "issue": [
    {
      "severity": "error",
      "code": "processing",
      "diagnostics": "The extension http://hl7.org/fhir/StructureDefinition/elementdefinition-identifier is not allowed to be used at this point (allowed = e:ElementDefinition; this element is [[Patient])",
      "location": [
        "Patient",
        "Line 1, Col 2"
      ]
    }
  ]
}
```
**Expected behavior:**
There is no error and no warning in the result.

**Environment:**
 - hapi-fhir v5.0.1
 - hapi-fhir-jpaserver-starter v5.0.0 with parent/version in pom.xml modified to 5.0.1
 - fhir_version=R4

**Additional context**
 - This is a regression as the JPA Server 4.2.0 do not have the problem.

 - The same regression is found on the public HAPI server (which is at 5.1.0-SNAPSHOT version)
   http://hapi.fhir.org/baseR4

 - The same problem happen with custom extensions created from profiles.

 - Strangely, the validation is **successful** when using hapi-fhir-cli:
`java -jar hapi-fhir-cli.jar validate --file patient.json --fhir-version R4`

**Comment by jamesagnew** [comment]
The [definition](http://hl7.org/fhir/extension-elementdefinition-identifier.xml.html) for that extension states that it has a context of `ElementDefinition` meaning that it is only valid if it is used specifically as an extension on an object of type ElementDefinition, and in your example you are using it on an object of type Identifier.

So the validator is correct in flagging this as an error. Previous versions of the validator did not validate extensions for having the correct context, which is why this wasn't flagged in previous versions of HAPI.

**Comment by j-fbriere** [comment]
Thank you very much for the answer. 
I had indeed a bad understanding of context in the case of an extension constraint.
For our custom profiles (StructureDefinition) with Extension constraint, we were using context.type = extension.
We adjusted them appropriately to context.type = element and now all of our tests pass.
And for the HL7 elementdefinition-identifier, since it can't apply to our usage, we created another custom StructureDefinition that meets our need.
Thanks again!
